Fixed race condition in LoggingOutputStream

dmatej opened the pull request:
* Closes #25438 
* Test is now reproducible (repeated 100 times, failed 3-40% locally, depending on CPU profile)
* Few more small improvements

dmatej committed:
LoggingPrintStream is now reproducible for race conditions

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Fixed randomly losing few records on LoggingOutputStream's shutdown

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
LogCollectorHandler - throw ISE rather than adding null to list

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Increased initial StringBuilder capacity in format methods.

- When using the format method the final string will be very probably larger
  than original template.

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

pzygielo commented on nucleus/glassfish-jul-extension/src/main/java/org/glassfish/main/jul/handler/LogCollectorHandler.java:
```diff
@@ -106,7 +106,12 @@ public GlassFishLogRecord pop() {
-            list.add(this.buffer.poll());
+            final GlassFishLogRecord record = this.buffer.poll();
+            if (record == null) {
+                // Unreachable code, but just in case.
```
To help me understand. This can be reached if after checking `!this.buffer.isEmpty()` here, but before `this.buffer.poll()` some other party empties `buffer`.
`this.buffer` is a field in this class, that could be MT-access protected, but it's decided to not do that.
And if this pretty normal situation happens, suddenly `ISEx` is thrown.
:eyes:

dmatej replied:
This class is used usually just in tests to catch logs and process them at some safe point, so null means that we do the check too early because logs are still coming.
Original code would just pass take what was already there, this code should "panic" and report at least some flaky test.

chabala commented on nucleus/glassfish-jul-extension/src/main/java/org/glassfish/main/jul/handler/LoggingOutputStream.java:
```diff
@@ -165,14 +165,25 @@ public void run() {
+                this.interrupt();
+            }
+            try {
+                this.join();
```
What's going on with all the unnecessary `this` in this class?

dmatej replied:
Sometimes it helps tired devs to think easily ;-)

And in one scenario it is very useful - when refactoring fields it prevents mistake collisions (usually reported by checkstyle as warnings, also by Eclipse IDE, we have lot of such cases in GF sources).

But probably just like you, I would not usually use `this` for methods. Seems, when I was writing this code I was thinking about which thread is waiting for the other. The calling thread is waiting for the end of the called, owning the method.

The more important on this code is previous wrong assumption that when I interrupt a thread, it is immediately interrupted and I don't need to wait for its end. I do wonder how many times I did the same silly mistake.

Currently we don't have any checkstyle rule forbidding unnecessary `this`.

dmatej committed:
Reflected reviews - reduced redundancy

- removed redundant this for methods
- simplified iterations with nullchecks

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Fixed SshClusterITest - sshd attack prevention

- `sshd re-exec requires execution with an absolute path` error message
- See https://www.openssh.com/releasenotes.html
- I did not find the cause of this change in behavior, just old explanations

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>
